dlib/test/constexpr_if.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+
+#include <string>
+#include <dlib/constexpr_if.h>
+#include <dlib/invoke.h>
+#include "tester.h"
+
+namespace
+{
+    using namespace test;
+    using namespace dlib;
+
+    logger dlog("test.constexpr_if");
+
+    struct A
+    {
+        int i;
+    };
+
+    struct B
+    {
+        float f;
+    };
+
+    struct C
+    {
+        std::string str;
+    };
+
+    template<typename T>
+    auto handle_type_and_return1(T obj)
+    {
+        return switch_type_<T>(
```

Comment:
So in C++14 you can totally do stuff like this
```
switch(some_object,
    [](int) { cout << "some_object is an int type"; },
    [](foo) { cout << "some_object is a foo type"; },
    [](double) { cout << "some_object is a double type"; },
    [](auto) { cout << "some_object is some other type!"; },
);
```
I've written stuff like that and seen it elsewhere.  How about doing that instead?

Reply:
What if you want to access a member function depending on the type ? Basically, whatever argument you have for if constexpr() applies here. Hopefully the tests show this. I started applying it to the DNN stuff in places where if constexpr would have been useful

Reply:
I think your method won't work if you want each case to return something different and then the calling function to return different things

Reply:
Also in your case all the branches are compiled. In this code, they aren't. Exactly like it constexpr()

Reply:
Yeah they all get compiled this way.  I don't know if that ends up mattering though in most cases :shrug:   But yeah I could imagine cases where it adds to compile time.